### PR TITLE
Cloudfront website for proof results

### PR DESCRIPTION
**Pull request by eidelmanjonathan**
This will allow us to deploy a Cloudfront website that exposes the details of the proof results

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: bin/cbmc-batch [hunk 1]
```diff
@@ -33,11 +33,11 @@ def prepare_paths(opts, quiet=True):
             abort("Bucket does not exist: {}".format(bkt))
 
     if opts['copysrc']:
-        s3.sync_directory_to_bucket(opts['srcdir'], opts['srcbucket'], quiet)
+        s3.sync_directory_to_bucket(opts['srcdir'], opts['srcbucket'], quiet, metadata={"cbmc": "True"})
```
Comment (danielsn) [inline bin/cbmc-batch]: What does this mean?

File: bin/s3.py [hunk 2]
```diff
@@ -527,8 +531,16 @@ def sync_directory_to_bucket(directory, bucket, quiet=False, delete=False):
             cmd.append('--delete')
         if quiet:
             cmd.append('--quiet')
+        if metadata:
+            cmd.append('--metadata')
+            param_str = ""
+            for k in metadata.keys():
```
Comment (danielsn) [inline bin/s3.py]: is doing the explicit call to keys expensive?

File: bin/s3.py [hunk 3]
```diff
@@ -527,8 +531,16 @@ def sync_directory_to_bucket(directory, bucket, quiet=False, delete=False):
...
+                param_str += '{}={},'.format(k, metadata[k])
+            param_str = param_str[:len(param_str) - 1]
```
Comment (danielsn) [inline bin/s3.py]: I guess this strips the final comma.  Is this a common idiom?

File: bin/s3.py [hunk 4]
```diff
@@ -527,8 +531,16 @@ def sync_directory_to_bucket(directory, bucket, quiet=False, delete=False):
...
+            cmd.append(param_str)
         if not quiet:
             print("Copying directory {} to bucket {}".format(directory, url))
+            print("Running copy command: {}".format(cmd))
```
Comment (danielsn) [inline bin/s3.py]: `print` or `log`

File: ci/snapshot/cloudfront.yaml [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+  TheBucketPolicy2:
+    Type: AWS::S3::BucketPolicy
+    Properties:
+      Bucket: !Ref S3BucketProofs
+      PolicyDocument:
+        Statement:
+          - Action:
+              - s3:GetObject
+            Effect: Allow
+            Resource: !Join [ "", [ "arn:aws:s3:::", !Ref S3BucketProofs, "/*/out/*" ] ]
```
Comment (danielsn) [inline ci/snapshot/cloudfront.yaml]: The double `*` worries me.  Is there any better way to do this?

File: ci/snapshot/cloudfront.yaml [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        Timeout: 5
+#        AutoPublishAlias: live
+        Code:
+          ZipFile: !Sub |
+            import boto3
+            PROOFS_BKT = "${S3BucketProofs}"
+            def lambda_handler(event, context):
+              request = event['Records'][0]['cf']['request']
+              headers = request['headers']
+              print(event)
```
Comment (danielsn) [inline ci/snapshot/cloudfront.yaml]: `log`?

File: ci/snapshot/cloudfront.yaml [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+              s3_filename = request['uri']
+              print(s3_filename)
```
Comment (danielsn) [inline ci/snapshot/cloudfront.yaml]: would it make sense to just log the whole request

File: ci/snapshot/cloudfront.yaml [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+              if "Metadata" in s3_obj.keys() and  "cbmc" in s3_obj['Metadata'].keys() and  s3_obj['Metadata']["cbmc"] == "True":
+                print("This file is good!")
+              else:
+                raise Exception("Tried to access bad file!")
+              headers['strict-transport-security'] = [{'key': 'Strict-Transport-Security', 'value': 'max-age=63072000; includeSubdomains; preload'}]
+              headers['content-security-policy'] = [{'key': 'Content-Security-Policy', 'value': "default-src 'none'; img-src 'self'; script-src 'self'; style-src 'self'; object-src 'none'"}]
+              headers['x-content-type-options'] = [{'key': 'X-Content-Type-Options', 'value': 'nosniff'}]
+              headers['x-frame-options'] = [{'key': 'X-Frame-Options', 'value': 'DENY'}]
+              headers['x-xss-protection'] = [{'key': 'X-XSS-Protection', 'value': '1; mode=block'}]
+              headers['referrer-policy'] = [{'key': 'Referrer-Policy', 'value': 'same-origin'}]
```
Comment (danielsn) [inline ci/snapshot/cloudfront.yaml]: Where do these values come from

File: ci/snapshot/deployment_tools/account_orchestration/AccountOrchestrator.py [hunk 9]
```diff
@@ -1,12 +1,14 @@
...
+from deployment_tools.account_orchestration.stacks_data import GLOBALS_CLOUDFORMATION_DATA, \
+    BUILD_TOOLS_CLOUDFORMATION_DATA, \
     PROOF_ACCOUNT_GITHUB_CLOUDFORMATION_DATA, BUILD_TOOLS_BUCKET_POLICY, PROOF_ACCOUNT_BATCH_CLOUDFORMATION_DATA, \
-    BUILD_TOOLS_PACKAGES, PROOF_ACCOUNT_PACKAGES, BUILD_TOOLS_ALARMS
+    BUILD_TOOLS_PACKAGES, PROOF_ACCOUNT_PACKAGES, BUILD_TOOLS_ALARMS, CLOUDFRONT_CLOUDFORMATION_DATA
 from deployment_tools.account_orchestration.AwsAccount import AwsAccount
 from deployment_tools.aws_managers.key_constants import BUILD_TOOLS_SNAPSHOT_ID_KEY, \
-    BUILD_TOOLS_ACCOUNT_ID_OVERRIDE_KEY, PROOF_ACCOUNT_ID_TO_ADD_KEY, PIPELINES_KEY, SNAPSHOT_ID_OVERRIDE_KEY
+    BUILD_TOOLS_ACCOUNT_ID_OVERRIDE_KEY, PROOF_ACCOUNT_ID_TO_ADD_KEY, PIPELINES_KEY, SNAPSHOT_ID_OVERRIDE_KEY, \
+    S3_BUCKET_PROOFS_OVERRIDE_KEY, CLOUDFRONT_URL_KEY
```
Comment (danielsn) [inline ci/snapshot/deployment_tools/account_orchestration/AccountOrchestrator.py]: alphabatize

File: bin/cbmc-batch (hunk 1 above)
Comment (eidelmanjonathan) [reply bin/cbmc-batch]: As a security precaution we want to make sure we only have cloudfront serve files that we know are related to the proof. So I basically add some S3 metadata to all of the files we're uploading to let Cloudfront know to serve those files. I'll clarify this in the code

File: bin/s3.py (hunk 2 above)
Comment (eidelmanjonathan) [reply bin/s3.py]: Fixed

File: bin/s3.py (hunk 3 above)
Comment (eidelmanjonathan) [reply bin/s3.py]: Replaced this with something more standard

File: bin/s3.py (hunk 4 above)
Comment (eidelmanjonathan) [reply bin/s3.py]: The whole file uses print and I didn't want to bloat this PR. In the future maybe we want to switch everything to use a logger

File: ci/snapshot/cloudfront.yaml (hunk 5 above)
Comment (eidelmanjonathan) [reply ci/snapshot/cloudfront.yaml]: I'm not sure it makes a difference. Any pattern we use will be public so if an adversary with write access to the S3 bucket wants to put something there, they can. What we're trying to defend against is silly mistakes by people with write access to the bucket. We do this by having metadata that tells Cloudfront the file was generated by our CI, and we don't serve anything without that metadata

File: ci/snapshot/cloudfront.yaml (hunk 6 above)
Comment (eidelmanjonathan) [reply ci/snapshot/cloudfront.yaml]: Same as above

File: ci/snapshot/cloudfront.yaml (hunk 7 above)
Comment (eidelmanjonathan) [reply ci/snapshot/cloudfront.yaml]: Yes. Will fix

File: ci/snapshot/cloudfront.yaml (hunk 8 above)
Comment (eidelmanjonathan) [reply ci/snapshot/cloudfront.yaml]: I basically added security headers based on a bunch of online tutorials like this one https://www.keycdn.com/blog/http-security-headers

File: ci/snapshot/deployment_tools/account_orchestration/AccountOrchestrator.py (hunk 9 above)
Comment (eidelmanjonathan) [reply ci/snapshot/deployment_tools/account_orchestration/AccountOrchestrator.py]: done

File: ci/snapshot/cloudfront.yaml [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+#        AutoPublishAlias: live
+        Code:
+          ZipFile: !Sub |
+            import boto3
+            import logging
+            logger = logging.getLogger('CloudFrontLambdaEdge')
+            logger.setLevel(logging.INFO)
+            PROOFS_BKT = "${S3BucketProofs}"
+            def lambda_handler(event, context):
+              request = event['Records'][0]['cf']['request']
```
Comment (danielsn) [inline ci/snapshot/cloudfront.yaml]: This code doesn't seem to do much error handling. Is that typical? We just let it crash and look in the logs?

File: ci/snapshot/cloudfront.yaml [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+              headers['x-frame-options'] = [{'key': 'X-Frame-Options', 'value': 'DENY'}]
+              headers['x-xss-protection'] = [{'key': 'X-XSS-Protection', 'value': '1; mode=block'}]
+              headers['referrer-policy'] = [{'key': 'Referrer-Policy', 'value': 'same-origin'}]
+              return request
+  IndexLambdaVersion:
+    Type: Custom::LatestLambdaVersion
+    Properties:
+      ServiceToken: !GetAtt PublishLambdaVersion.Arn
+      FunctionName: !Ref LambdaEdgeFunctionSample
+#      Nonce: !Ref Nonce
```
Comment (danielsn) [inline ci/snapshot/cloudfront.yaml]: why commented?

File: ci/snapshot/deployment_tools/account_orchestration/AccountOrchestrator.py [hunk 12]
```diff
@@ -183,6 +193,16 @@ def deploy_proof_account_stacks(self):
                                              BUILD_TOOLS_ACCOUNT_ID_OVERRIDE_KEY: self.build_tools.account_id
                                          })
 
+    def deploy_cloudfront_stacks(self):
+        self.cloudfront_account.deploy_stacks(CLOUDFRONT_CLOUDFORMATION_DATA,
```
Comment (danielsn) [inline ci/snapshot/deployment_tools/account_orchestration/AccountOrchestrator.py]: mix of named and positional prams

File: ci/snapshot/deployment_tools/account_orchestration/stacks_data.py [hunk 13]
```diff
@@ -61,6 +61,14 @@
     "template": {"extract": True}
 }
 
+
+CLOUDFRONT_CLOUDFORMATION_DATA = {
+    "proof-results-cloudfront": {
+        TEMPLATE_NAME_KEY: "cloudfront.yaml",
+        PARAMETER_KEYS_KEY: ['S3BucketProofs']
```
Comment (danielsn) [inline ci/snapshot/deployment_tools/account_orchestration/stacks_data.py]: alphabatize

File: ci/snapshot/deployment_tools/account_orchestration/stacks_data.py [hunk 14]
```diff
@@ -102,6 +110,7 @@
                              'GitHubBranchName',
                              'GitHubRepository',
                              'ProjectName',
+                             'CloudfrontUrl',
                              'S3BucketToolsName',
                              'SnapshotID']
```
Comment (danielsn) [inline ci/snapshot/deployment_tools/account_orchestration/stacks_data.py]: alphabatize

File: ci/snapshot/deployment_tools/aws_managers/key_constants.py [hunk 15]
```diff
@@ -1,3 +1,5 @@
+CLOUDFRONT_URL_KEY = "CloudfrontUrl"
+S3_BUCKET_PROOFS_OVERRIDE_KEY = "S3BucketProofs"
 TEMPLATE_NAME_KEY = "template_name"
 PARAMETER_KEYS_KEY = "parameter_keys"
 PARAMETER_OVERRIDES_KEY = "parameter_overrides"
```
Comment (danielsn) [inline ci/snapshot/deployment_tools/aws_managers/key_constants.py]: alphabatize

File: bin/cbmc-batch (hunk 1 above)
Comment (eidelmanjonathan) [reply bin/cbmc-batch]: Done

File: bin/s3.py [hunk 16]
```diff
@@ -527,8 +527,16 @@ def sync_directory_to_bucket(directory, bucket, quiet=False, delete=False):
             cmd.append('--delete')
         if quiet:
             cmd.append('--quiet')
+        if metadata:
+            cmd.append('--metadata')
+            param_str = ""
+            for k in metadata:
+                param_str += '{}={},'.format(k, metadata[k])
```
Comment (danielsn) [inline bin/s3.py]: f string

File: bin/s3.py (hunk 4 above)
Comment (danielsn) [reply bin/s3.py]: Do we have a backlog item for that?

File: ci/snapshot/cbmc_ci_end.py [hunk 17]
```diff
@@ -147,6 +151,12 @@ def lambda_handler(event, context):
                         response['status'] = clog_writert.FAILED
                 else:
                     response['status'] = clog_writert.FAILED
+            elif job_name_info.is_cbmc_report_job():
+                print("type: {}, is_cbmc_report_job: {}, job name: {}".format(job_name_info.type,
```
Comment (danielsn) [inline ci/snapshot/cbmc_ci_end.py]: f string

File: ci/snapshot/cloudfront.yaml (hunk 5 above)
Comment (danielsn) [reply ci/snapshot/cloudfront.yaml]: OK

File: ci/snapshot/cloudfront.yaml [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+      Properties:
+        Role: !GetAtt LambdaEdgeFunctionRole.Arn
+        Runtime: python3.7
+        Handler: index.lambda_handler
+        Timeout: 5
+        Code:
+          ZipFile: !Sub |
+            import boto3
+            import logging
+            logger = logging.getLogger('CloudFrontLambdaEdge')
```
Comment (danielsn) [inline ci/snapshot/cloudfront.yaml]: Is this standard? Instead of `logging.getLogger(__name__)`

File: ci/snapshot/cloudfront.yaml (hunk 10 above)
Comment (danielsn) [reply ci/snapshot/cloudfront.yaml]: OK

**Comment by danielsn** [comment]
Should the .pyc files really be included in the commit

File: ci/snapshot/deployment_tools/account_orchestration/AccountOrchestrator.py [hunk 19]
```diff
@@ -128,23 +137,26 @@ def add_proof_account_to_shared_bucket_policy(self):
                                        overrides=param_overrides)
 
     #
-    def deploy_proof_account_github(self):
+    def deploy_proof_account_github(self, cloudfront_url=None):
         """
         Deploys the 'github' stack in the proof account
+        cloudfront_url: string - What to post as the URL to our Cloudfront server for proof details
         """
         self.logger.info("Deploying github stack in proof account {}".format(self.proof_account.account_id))
```
Comment (danielsn) [inline ci/snapshot/deployment_tools/account_orchestration/AccountOrchestrator.py]: f string

File: bin/cbmc-batch [hunk 20]
```diff
@@ -31,13 +32,14 @@ def prepare_paths(opts, quiet=True):
         bkt = s3.bucket_name(path)
         if not s3.bucket_exists(bkt, region=opts['region']):
             abort("Bucket does not exist: {}".format(bkt))
```
Comment (danielsn) [inline bin/cbmc-batch]: f string

File: bin/s3.py [hunk 21]
```diff
@@ -527,8 +527,16 @@ def sync_directory_to_bucket(directory, bucket, quiet=False, delete=False):
...
+            param_str = param_str.rstrip(",")
+            cmd.append(param_str)
         if not quiet:
             print("Copying directory {} to bucket {}".format(directory, url))
```
Comment (danielsn) [inline bin/s3.py]: f string

File: ci/snapshot/cloudfront.yaml [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+            PROOFS_BKT = "${S3BucketProofs}"
+            def lambda_handler(event, context):
+              request = event['Records'][0]['cf']['request']
+              headers = request['headers']
+              s3_filename = request['uri']
+              s3_obj = boto3.client("s3").get_object(Bucket=PROOFS_BKT, Key=s3_filename[1:])
+              if "Metadata" in s3_obj.keys() and  "public-website-contents" in s3_obj['Metadata'] and  s3_obj['Metadata']["public-website-contents"] == "True":
+                logger.info(f"Requested file {s3_filename} has the required metadata. Serving file...")
+              else:
+                raise Exception("Tried to access bad file {s3_filename} that has no metadata!")
```
Comment (danielsn) [inline ci/snapshot/cloudfront.yaml]: I think you need a f here

File: ci/snapshot/cloudfront.yaml (hunk 8 above)
Comment (danielsn) [reply ci/snapshot/cloudfront.yaml]: Is this something we will want in more than one place, in which case it should be a utility function

File: bin/s3.py (hunk 4 above)
Comment (eidelmanjonathan) [reply bin/s3.py]: Not for this specific issue but for generally refactoring this script

File: ci/snapshot/cbmc_ci_end.py (hunk 17 above)
Comment (eidelmanjonathan) [reply ci/snapshot/cbmc_ci_end.py]: Fixed

File: ci/snapshot/cloudfront.yaml (hunk 18 above)
Comment (eidelmanjonathan) [reply ci/snapshot/cloudfront.yaml]: I feel like that's not a good idea in inline code. What will the module be with __name__?

File: ci/snapshot/deployment_tools/account_orchestration/AccountOrchestrator.py (hunk 19 above)
Comment (eidelmanjonathan) [reply ci/snapshot/deployment_tools/account_orchestration/AccountOrchestrator.py]: done

File: ci/snapshot/cloudfront.yaml (hunk 22 above)
Comment (eidelmanjonathan) [reply ci/snapshot/cloudfront.yaml]: Fixed

**Pull request merged**